src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostInlineCompletionEndpoint.cs

```diff
@@ -0,0 +1,115 @@
+using Microsoft.VisualStudio.Razor.Settings;
+using Roslyn.LanguageServer.Protocol;
+using VSLSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+
+#pragma warning disable RS0030 // Do not use banned APIs
+[Shared]
+[CohostEndpoint(VSInternalMethods.TextDocumentInlineCompletionName)]
+[Export(typeof(IDynamicRegistrationProvider))]
+[ExportCohostStatelessLspService(typeof(CohostInlineCompletionEndpoint))]
+[method: ImportingConstructor]
+#pragma warning restore RS0030 // Do not use banned APIs
+internal sealed class CohostInlineCompletionEndpoint(
+    IRemoteServiceInvoker remoteServiceInvoker,
+    IClientSettingsManager clientSettingsManager)
+    : AbstractRazorCohostDocumentRequestHandler<VSInternalInlineCompletionRequest, VSInternalInlineCompletionList?>, IDynamicRegistrationProvider
+{
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
+    private readonly IClientSettingsManager _clientSettingsManager = clientSettingsManager;
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    public ImmutableArray<VSLSP.Registration> GetRegistrations(VSLSP.VSInternalClientCapabilities clientCapabilities, RazorCohostRequestContext requestContext)
+    {
+        if (clientCapabilities.TextDocument?.CodeAction?.DynamicRegistration == true)
+        {
+            return [new VSLSP.Registration
+            {
+                Method = VSInternalMethods.TextDocumentInlineCompletionName,
+                RegisterOptions = new VSLSP.VSInternalInlineCompletionRegistrationOptions().EnableInlineCompletion()
+            }];
+        }
+
+        return [];
+    }
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(VSInternalInlineCompletionRequest request)
+        => request.TextDocument.ToRazorTextDocumentIdentifier();
+
+    protected override Task<VSInternalInlineCompletionList?> HandleRequestAsync(VSInternalInlineCompletionRequest request, RazorCohostRequestContext context, CancellationToken cancellationToken)
+        => HandleRequestAsync(context, context.TextDocument.AssumeNotNull(), request.Position.ToLinePosition(), request.Options, cancellationToken);
+
+    private async Task<VSInternalInlineCompletionList?> HandleRequestAsync(RazorCohostRequestContext? context, TextDocument razorDocument, LinePosition position, FormattingOptions formattingOptions, CancellationToken cancellationToken)
+    {
+        var requestInfo = await _remoteServiceInvoker.TryInvokeAsync<IRemoteInlineCompletionService, InlineCompletionRequestInfo?>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetInlineCompletionInfoAsync(solutionInfo, razorDocument.Id, position, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (requestInfo is not { } info)
```

Comment:
💡 Since this is a record struct, you could deconstruct is in the pattern match:

```suggestion
        if (requestInfo is not InlineCompletionRequestInfo(var generatedDocumentUri, var position))
```

Then you can use the values below.

Reply:
ugh, thank you! I was fighting with `if (requestInfo is not { } (var generatedDocumentUri, var position))` trying to remember the syntax, so I just gave up 🤦‍♂️

Deconstruct patterns look too much like method calls for my poor brain to remember them :)